File: lab/python/matematica_e_operadores/torre_exponenciacao.py
from math import gcd

# --- potências e auxiliares ---------------------------------


def powmod(a, e, m):
    a %= m
    r = 1
    while e > 0:
        if e & 1:
            r = (r * a) % m
        a = (a * a) % m
        e >>= 1
    return r


def geq_pow_threshold(b, c, T):
    """Retorna True se b^c >= T, sem calcular b^c por completo.
       Para T<=1, retorna True (qualquer b^c >=1)."""
    if T <= 1:
        return True
    if c == 0:
        return 1 >= T
    # expo rápido com “cap” em T
    res = 1
    base = b
    exp = c
    while exp > 0 and res < T:
        if exp & 1:
            res *= base
            if res >= T:
                return True
        exp >>= 1
        if exp > 0:
            # evita crescer demais: se base já ultrapassa sqrt(T),
            # próximo quadrado certamente excede T
            if base >= T:
                return True
            base_sq = base * base
            if base_sq >= T and exp:  # ainda haverá uso
                return True
            base = base_sq
    return res >= T


# --- fatoração e funções aritméticas -------------------------

def factor(n):
    """Fatoração por divisão tentativa (suficiente para exercício)."""
    f = []
    d = 2
    while d * d <= n:
        if n % d == 0:
            k = 0
            while n % d == 0:
                n //= d
                k += 1
            f.append((d, k))
        d += 1 if d == 2 else 2  # 2 e depois ímpares
    if n > 1:
        f.append((n, 1))
    return f


def vp(a, p):
    """Expoente de p em a (v_p)."""
    t = 0
    while a % p == 0:
        a //= p
        t += 1
    return t, a


def phi_prime_power(p, k):
    if k == 0:
        return 1
    return (p - 1) * (p ** (k - 1))


def lambda_prime_power(p, k):
    # Carmichael:
    # - p ímpar: λ(p^k) = φ(p^k)
    # - p = 2: λ(2)=1, λ(4)=2, λ(2^k)=2^{k-2} para k>=3
    if p != 2:
        return phi_prime_power(p, k)
    if k == 1:
        return 1
    if k == 2:
        return 2
    return 1 << (k - 2)


def lcm(a, b):
    return a // gcd(a, b) * b


def carmichael_lambda(n):
    """λ(n) por fatoração em potências primas e MMC das λ(p^k)."""
    lam = 1
    for p, k in factor(n):
        lam = lcm(lam, lambda_prime_power(p, k))
    return lam


# --- CRT para módulos coprimos -------------------------------

def invmod(a, m):
    # extensão de Euclides (m e a coprimos)
    r0, r1 = a % m, m
    s0, s1 = 1, 0
    while r1:
        q = r0 // r1
        r0, r1 = r1, r0 - q * r1
        s0, s1 = s1, s0 - q * s1
    if r0 != 1:
        raise ValueError("inv inexistente")
    return s0 % m


def crt_pair(a1, m1, a2, m2):
    """Resolve x≡a1 (mod m1), x≡a2 (mod m2), com m1,m2 coprimos."""
    inv = invmod(m1, m2)
    k = (a2 - a1) * inv % m2
    x = a1 + m1 * k
    return x, m1 * m2


# --- núcleo por potência prima -------------------------------

def tower_mod_prime_power(a, b, c, p, k):
    mod = p ** k
    a %= mod
    if mod == 1:
        return 0

    t, a_coprime = vp(a if a != 0 else mod, p)  # v_p(a), a' coprimo a p
    # Se a==0 mod p^k, já é 0
    if t >= k:
        return 0

    # Precisaremos comparar E=b^c contra thresholds:
    # T1 = ceil(k/t) (para decidir se p^{tE} zera), se t>0
    # T2 = λ(p^k) (para redução do expoente na parte coprima)
    lam = lambda_prime_power(p, k)

    # Funções para E:
    def E_geq(T):
        return geq_pow_threshold(b, c, T)

    def E_mod_L_with_shift(L):
        # e_red = E mod L
        e_red = powmod(b, c, L)
        # se E >= L, somamos L para evitar expoente 0 em casos grandes
        if E_geq(L):
            e_red += L
        return e_red

    if t == 0:
        # coprimos com p^k
        e = E_mod_L_with_shift(lam)
        return powmod(a, e, mod)

    # t>0: separa a = p^t * a_coprime
    T1 = (k + t - 1) // t  # ceil(k/t)
    if E_geq(T1):
        return 0  # p^{tE} >= p^k

    # senão, calculamos p^{tE} * (a')^E mod p^k
    # precisamos de E "inteiro" em dois lugares:
    # - potência de p: t*E (aqui basta saber valor exato até ser pequeno)
    # - expoente de a_coprime reduzido por lam (com shift)
    # E pode ser pequeno (já que não atingiu T1). Vamos obtê-lo de forma exata,
    # mas com early stop: como T1 é o limite,
    # calcular b^c exatamente (é pequeno).
    # Para segurança, construímos por exponenciação rápida sem módulo.
    def exact_pow(b, c):
        r, base, exp = 1, b, c
        while exp > 0:
            if exp & 1:
                r *= base
            base *= base
            exp >>= 1
        return r

    E = exact_pow(b, c)  # seguro pois E < T1 e T1 <= ~k/t
    p_part = powmod(p, t * E, mod)
    e2 = E_mod_L_with_shift(lam)  # redução segura p/ a' coprimo
    a_part = powmod(a_coprime % mod, e2, mod)
    return (p_part * a_part) % mod


# --- função principal ----------------------------------------

def tower_mod(a, b, c, m):
    if m == 1:
        return 0
    # fatora m e resolve por CRT
    residues = []
    moduli = []
    for p, k in factor(m):
        r = tower_mod_prime_power(a, b, c, p, k)
        residues.append(r)
        moduli.append(p ** k)

    # junta com CRT
    x, M = residues[0], moduli[0]
    for r, mod in zip(residues[1:], moduli[1:]):
        x, M = crt_pair(x, M, r, mod)
    return x % m


# ----------------- exemplos rápidos --------------------------
if __name__ == "__main__":
    # Exemplos clássicos
    print(tower_mod(2, 3, 4, 1000))     # 2^(81) mod 1000 -> 352
    print(tower_mod(7, 13, 5, 97))      # (coprimos) usa λ(97)=96
    print(tower_mod(12, 5, 3, 2**10))   # m potência de 2 e a múltiplo de 2

File: lab/javascript/matematica_e_operadores/torre_exponenciacao.js
// torre.js

// ---------- helpers BigInt ----------

const ZERO=0n, ONE=1n;

function gcd(a,b){ while(b){ [a,b]=[b,a%b]; } return a; }

function mulmod(a,b,m){ return (a*b) % m; }

function powmod(a,e,m){
  if (m===ONE) return ZERO;
  a%=m;
  let r=ONE;
  while (e>ZERO){
    if (e & ONE) r = (r*a) % m;
    a = (a*a) % m;
    e >>= ONE;
  }
  return r;
}

function vp(a,p){
  let t=0n;
  if (p===ONE) return {t:0n, aCoprimo:a};
  while (a % p === ZERO){ a/=p; t++; }
  return {t, aCoprimo:a};
}

function phiPrimePower(p,k){
  if (k===0n) return ONE;
  let pk=ONE;
  for (let i=0n;i<k;i++) pk*=p;
  return (p-ONE) * (pk/p);
}

function lambdaPrimePower(p,k){
  if (p!==2n) return phiPrimePower(p,k);
  if (k===1n) return 1n;
  if (k===2n) return 2n;
  return 1n << (k-2n);
}

function invmod(a,m){
  // ext. Euclides
  let r0 = ((a%m)+m)%m, r1=m, s0=1n, s1=0n;
  while (r1){
    const q = r0 / r1;
    [r0,r1] = [r1, r0 - q*r1];
    [s0,s1] = [s1, s0 - q*s1];
  }
  if (r0 !== 1n && r0 !== -1n) throw new Error("inv inexistente");
  if (r0 === -1n) s0 = -s0;
  s0 %= m; if (s0 < 0n) s0 += m;
  return s0;
}

function crtPair(a1,m1,a2,m2){
  const inv = invmod(m1, m2);
  let k = (a2 - a1) % m2; if (k<0n) k+=m2;
  k = (k * inv) % m2;
  const x = (a1 + m1*k) % (m1*m2);
  return [x, m1*m2];
}

function pk(p,k){
  let x=ONE; for (let i=0n;i<k;i++) x*=p; return x;
}

// b^c >= T ?  (T>=1)
function geqPowThreshold(b,c,T){
  if (T<=ONE) return true;
  if (c===ZERO) return 1n >= T;
  let res=ONE, base=b, exp=c;
  while (exp && res < T){
    if (exp & ONE){
      res *= base;
      if (res >= T) return true;
    }
    exp >>= ONE;
    if (exp){
      base = base*base;
      if (base >= T) return true;
    }
  }
  return res >= T;
}

function E_mod_L_with_shift(b,c,L){
  if (L===ONE) return ONE; // evita expoente 0
  let eRed = powmod(b,c,L);
  if (geqPowThreshold(b,c,L)) eRed += L;
  return eRed;
}

function exactPowBounded(b,c,limit){
  // usado apenas quando sabemos que b^c < limit
  let r=ONE, base=b, exp=c;
  while (exp){
    if (exp & ONE) r *= base;
    exp >>= ONE;
    if (exp) base *= base;
  }
  return r; // < limit
}

// ---------- núcleo p^k ----------

function towerModPrimePower(a,b,c,p,k){
  const mod = pk(p,k);
  a %= mod;
  if (mod===ONE) return ZERO;

  const {t, aCoprimo} = vp(a ? a : mod, p);
  if (t >= k) return ZERO; // a ≡ 0 (mod p^k)

  const lam = lambdaPrimePower(p,k);

  if (t===0n){
    const e = E_mod_L_with_shift(b,c,lam);
    return powmod(a,e,mod);
  }

  const T1 = (k + t - 1n) / t; // ceil(k/t)
  if (geqPowThreshold(b,c,T1)) return ZERO;

  const E = exactPowBounded(b,c,T1);
  const pPart = powmod(p, t*E, mod);
  const e2 = E_mod_L_with_shift(b,c,lam);
  const aPart = powmod(aCoprimo % mod, e2, mod);
  return (pPart * aPart) % mod;
}

// ---------- principal ----------

function factorBig(n){
  const P=[], K=[];
  let d=2n;
  while (d*d<=n){
    if (n % d === 0n){
      let k=0n;
      while (n % d === 0n){ n/=d; k++; }
      P.push(d); K.push(k);
    }
    d += (d===2n)?1n:2n;
  }
  if (n>1n){ P.push(n); K.push(1n); }
  return [P,K];
}

function towerMod(a,b,c,m){
  if (m===ONE) return ZERO;
  const [P,K] = factorBig(m);

  let x=0n, M=1n;
  for (let i=0;i<P.length;i++){
    const p=P[i], k=K[i];
    const mi = pk(p,k);
    const ri = towerModPrimePower(a,b,c,p,k);
    if (M===1n){ x=ri; M=mi; }
    else{ [x,M] = crtPair(x,M,ri,mi); }
  }
  return ((x % m) + m) % m;
}

// --------- wrappers práticos ---------

const asBig = n => (typeof n === 'bigint' ? n : BigInt(n));

function towerModNum(a,b,c,m){
  return towerMod(asBig(a), asBig(b), asBig(c), asBig(m));
}

// --------- demo ---------

console.log(String(towerModNum(2,3,4,1000)));       // 352
console.log(String(towerModNum(7,13,5,97)));        // usa λ(97)=96
console.log(String(towerModNum(12,5,3, 1n<<10n)));  // módulo 2^10
